src/v/utils/bottomless_token_bucket.h

```diff
@@ -0,0 +1,105 @@
+inline int64_t
+muldiv(const int64_t x, const int64_t mul, const int64_t div) noexcept {
+    const auto qr = std::div(x, div);
+    return qr.quot * mul + qr.rem * mul / div;
+}
+
+/// Variation of token bucket algorithm to allow bursty traffic.
+/// Tokens represent resource units (e.g. bytes of traffic) whose usage is
+/// controlled. Number of tokens in the bucket is capped by \ref quota *
+/// \ref width at the top, and can go negative without a bound.
+class bottomless_token_bucket {
+public:
+    /// Quota is measured in [tokens/s] and cannot be negative despite the
+    /// signed type
+    using quota_t = int64_t;  // [tokens/s]
+    using tokens_t = int64_t; // [tokens]
+    using clock = ss::lowres_clock;
+    using time_res_t = std::chrono::milliseconds;
+
+    // Algoritms rely on time_res_t to be subsecond
+    static_assert(time_res_t::period::num == 1);
+
+    // Maximum supported width of a bucket ~ 25 days
+    static constexpr time_res_t max_width{std::numeric_limits<int32_t>::max()};
+
+    // To avoid dynamic checking of acceptable width and quota maximums,
+    // quota is limited in the way to avoid integer overflows in any condition.
+    // This effectively maxes it to ~ 2 Ttokens/s. If that is ever found not
+    // enough, this maximum should be raised to full int64_t, and both
+    // set_quota() and set_width() should dynamically check that the calculated
+    // burst does not overflow.
+    static constexpr quota_t max_quota{
+      static_cast<quota_t>(std::numeric_limits<int32_t>::max())
+      * time_res_t::period::den};
+
+    /// Token bucket is initialized full of tokens
+    bottomless_token_bucket(
+      const quota_t quota, const time_res_t width) noexcept {
+        // do boundary checks on params and calculate _burst
+        set_quota_impl(quota);
+        set_width_impl(width);
+        update_burst_tokens();
+        refill(clock::time_point{});
+    }
+
+    void set_quota(const quota_t quota) noexcept {
+        set_quota_impl(quota);
+        update_burst_tokens();
+    }
+
+    /// @post 0 < return value <= max_quota
+    quota_t quota() const noexcept { return _quota; }
+
+    void set_width(const time_res_t width) noexcept {
+        set_width_impl(width);
+        update_burst_tokens();
+    }
+
+    /// Record usage of tokens
+    /// @param n Number of tokens used
+    /// @param now Time to record at
+    void use(tokens_t n, clock::time_point now) noexcept;
+
+    /// Current amount of tokens
+    tokens_t tokens() const noexcept { return _tokens; }
+
+private:
+    void set_quota_impl(const quota_t quota) noexcept;
+    void set_width_impl(const time_res_t width) noexcept;
+    void update_burst_tokens() noexcept;
+    void refill(const clock::time_point now) noexcept;
+
+private:
+    quota_t _quota{0};
+    time_res_t _width;
+    tokens_t _tokens{0};
+    tokens_t _burst{0}; // capacity of the bucket
+    clock::time_point _last_check{-max_width};
```

Comment:
Other then the millisecond precision does this work the same as [this token bucket impl](https://github.com/redpanda-data/redpanda/blob/dev/src/v/kafka/server/token_bucket_rate_tracker.h) in our tree which allows the count to go negative as well? I'm thinking if the main differences are ms precision and dynamically adjustable parameters, we could just remove the existing one in favor of this new one here.

Reply:
There are more differences.

- token level can go negative unrestricted - because there are multiple
connections and a second connection could consume its initial portion
of tokens before the token have restored to zero after the first
connection has exhausted them below zero

- quota and bucket width can be changed - to allow dynamic adjustment of
the corresponding cluster settings

In theory we can merge these implementations and make time resolution and whether tokens can keep going to the negative many times into policy details, but I'm not sure if that really worth the effort given the simplicity of the token bucket algorithm, so maybe it's easier to keep two implementations and just document their distinction well.